File: python/collector_sdk/data_api_test.py
from __future__ import annotations

import json
import unittest

from collector_sdk.data_api import build_data_api_upsert_fn


class BuildDataAPIUpsertFnTests(unittest.TestCase):
    def test_posts_to_internal_dataset_upsert_endpoint(self) -> None:
        captured: dict[str, str] = {
            "url": "",
            "method": "",
            "body": "",
            "token": "",
        }

        def fake_request(url: str, method: str, headers: dict[str, str], body: bytes) -> tuple[int, str]:
            captured["url"] = url
            captured["method"] = method
            captured["token"] = headers.get("X-Ephemeral-Token", "")
            captured["body"] = body.decode("utf-8")
            return 200, '{"code":0}'

        upsert_fn = build_data_api_upsert_fn(
            capability_endpoint="http://chainbase-block-scanner-fetcher-test-svc",
            ephemeral_token="token-abc",
            request_fn=fake_request,
        )

        upsert_fn(
            {
                "run_id": "115",
                "task_name": "defillama_tvl_top10",
                "task_version": "v0.1.0",
                "attempt_no": 1,
                "dataset": "defillama_tvl_protocols_top10",
                "rows": [
                    {
                        "protocol": "Lido",
                        "rank": 1,
                        "tvl_usd": 123.4,
                        "snapshot_at": "2026-03-25T00:00:00Z",
                    }
                ],
                "key_fields": ["protocol"],
            }
        )

        self.assertEqual(
            captured["url"],
            "http://chainbase-block-scanner-fetcher-test-svc/api/internal/data/datasets/defillama_tvl_protocols_top10/upsert",
        )
        self.assertEqual(captured["method"], "POST")
        self.assertEqual(captured["token"], "token-abc")

        payload = json.loads(captured["body"])
        self.assertEqual(payload["key_fields"], ["protocol"])
        self.assertEqual(payload["audit"]["task_name"], "defillama_tvl_top10")
        self.assertEqual(payload["audit"]["task_version"], "v0.1.0")
        self.assertEqual(payload["audit"]["run_id"], 115)


if __name__ == "__main__":
    unittest.main()

File: typescript/src/data_api.test.ts
import assert from "node:assert/strict"
import { describe, it } from "node:test"

import { buildDataApiUpsertFn } from "./data_api"

describe("buildDataApiUpsertFn", () => {
  it("returns noop when capability endpoint is empty", async () => {
    let called = false
    const upsert = buildDataApiUpsertFn("   ", "token", async () => {
      called = true
      return { status: 200, text: "ok" }
    })

    await upsert({
      runId: "123",
      taskName: "demo_task",
      taskVersion: "v0.1.0",
      attemptNo: 1,
      dataset: "demo_dataset",
      rows: [{ id: "1" }],
      keyFields: ["id"],
    })

    assert.equal(called, false)
  })

  it("builds scanner upsert request with encoded dataset and audit run_id", async () => {
    const requests: Array<{ url: string; method: string; headers: Record<string, string>; body: string }> = []
    const upsert = buildDataApiUpsertFn(" https://scanner.internal/ ", "  ephemeral-token  ", async (url, method, headers, body) => {
      requests.push({ url, method, headers, body })
      return { status: 200, text: "ok" }
    })

    await upsert({
      runId: "42",
      taskName: "demo_task",
      taskVersion: "v0.1.0",
      attemptNo: 1,
      dataset: "space name/ok",
      rows: [{ id: "1", foo: "bar" }],
      keyFields: ["id"],
    })

    assert.equal(requests.length, 1)
    assert.deepEqual(requests[0], {
      url: "https://scanner.internal/api/internal/data/datasets/space%20name%2Fok/upsert",
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "X-Ephemeral-Token": "ephemeral-token",
      },
      body: JSON.stringify({
        rows: [{ id: "1", foo: "bar" }],
        key_fields: ["id"],
        audit: {
          task_name: "demo_task",
          task_version: "v0.1.0",
          run_id: 42,
        },
      }),
    })
  })

  it("omits audit run_id when runId is invalid and omits token when blank", async () => {
    const requests: Array<{ headers: Record<string, string>; body: string }> = []
    const upsert = buildDataApiUpsertFn("https://scanner.internal", "   ", async (_url, _method, headers, body) => {
      requests.push({ headers, body })
      return { status: 200, text: "ok" }
    })

    await upsert({
      runId: "run-abc",
      taskName: "demo_task",
      taskVersion: "v0.1.0",
      attemptNo: 1,
      dataset: "demo_dataset",
      rows: [{ id: "1" }],
      keyFields: ["id"],
    })

    assert.deepEqual(requests[0]?.headers, {
      "Content-Type": "application/json",
    })
    assert.deepEqual(JSON.parse(requests[0]?.body ?? "{}"), {
      rows: [{ id: "1" }],
      key_fields: ["id"],
      audit: {
        task_name: "demo_task",
        task_version: "v0.1.0",
      },
    })
  })

  it("throws when scanner API responds with non-2xx", async () => {
    const upsert = buildDataApiUpsertFn("https://scanner.internal", "token", async () => {
      return { status: 500, text: "boom" }
    })

    await assert.rejects(
      async () => {
        await upsert({
          runId: "123",
          taskName: "demo_task",
          taskVersion: "v0.1.0",
          attemptNo: 1,
          dataset: "demo_dataset",
          rows: [{ id: "1" }],
          keyFields: ["id"],
        })
      },
      /data upsert failed: http 500 boom/,
    )
  })
})
